src/test/AccessLockHook.sol

```diff
@@ -0,0 +1,196 @@
+        address, /* sender **/
+        PoolKey calldata key,
+        uint256, /* amount0 **/
+        uint256, /* amount1 **/
+        bytes calldata hookData
+    ) external override returns (bytes4) {
+        return _executeAction(key, hookData, IHooks.beforeDonate.selector);
+    }
+
+    function beforeModifyPosition(
+        address, /* sender **/
+        PoolKey calldata key,
+        IPoolManager.ModifyPositionParams calldata, /* params **/
+        bytes calldata hookData
+    ) external override returns (bytes4) {
+        return _executeAction(key, hookData, IHooks.beforeModifyPosition.selector);
+    }
+
+    function _executeAction(PoolKey memory key, bytes calldata hookData, bytes4 selector) internal returns (bytes4) {
+        if (hookData.length == 0) {
+            // We have re-entered the hook or we are initializing liquidity in the pool before testing the lock actions.
+            return selector;
+        }
+        (uint256 amount, LockAction action) = abi.decode(hookData, (uint256, LockAction));
+
+        // These actions just use some hardcoded parameters.
+        if (action == LockAction.Mint) {
+            manager.mint(key.currency1, address(this), amount);
+        } else if (action == LockAction.Take) {
+            manager.take(key.currency1, address(this), amount);
+        } else if (action == LockAction.Donate) {
+            manager.donate(key, amount, amount, new bytes(0));
+        } else if (action == LockAction.Swap) {
+            manager.swap(
+                key,
+                IPoolManager.SwapParams({
+                    zeroForOne: true,
+                    amountSpecified: int256(amount),
+                    sqrtPriceLimitX96: TickMath.MIN_SQRT_RATIO + 1
+                }),
+                new bytes(0)
+            );
+        } else if (action == LockAction.ModifyPosition) {
+            manager.modifyPosition(
+                key,
+                IPoolManager.ModifyPositionParams({tickLower: -60, tickUpper: 60, liquidityDelta: int256(amount)}),
+                new bytes(0)
+            );
+        } else if (action == LockAction.NoOp) {
+            assertEq(address(manager.getCurrentHook()), address(this));
+            return Hooks.NO_OP_SELECTOR;
+        } else {
+            revert("Invalid action");
+        }
+
+        return selector;
+    }
+}
+
+// Hook that can access the lock.
+// Also has the ability to call out to another hook or pool.
+contract AccessLockHook2 is Test, BaseTestHooks {
+    IPoolManager manager;
+
+    error IncorrectHookSet();
+
+    constructor(IPoolManager _manager) {
+        manager = _manager;
+    }
+
+    function beforeModifyPosition(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.ModifyPositionParams calldata params,
+        bytes calldata hookData
+    ) external override returns (bytes4) {
+        if (address(manager.getCurrentHook()) != address(this)) {
+            revert IncorrectHookSet();
+        }
```

Comment:
This seems like its testing a well behaved acess lock hook. Should we add a test hook that actively tries to access locks even though it's not the currentHook? There might already be tests for this that I might be missing too